### PR TITLE
Use a faster sort algorithm pdqsort for Array.Sort

### PR DESCRIPTION
**Pull request by hez2010**
Based on the work https://drops.dagstuhl.de/opus/volltexte/2016/6389/pdf/LIPIcs-ESA-2016-38.pdf and C++ implementation https://github.com/orlp/pdqsort.

I haven't implemented the branch-less partition right, which can be improved in the future.

Benchmark: https://gist.github.com/hez2010/6b52929ee1755788c34818972c46aefb

Fixes #68325

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs [hunk 1]
```diff
@@ -1103,6 +1908,28 @@ private static bool GreaterThan(ref TKey left, ref TKey right)
...
+            // return ref Unsafe.Add(ref MemoryMarshal.GetReference(span), index);
+            // TODO: investigate where can we use the above line with confidence instead of this.
+            return ref span[index];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static Span<T> UnguardedSlice<T>(Span<T> span, int begin, int end)
+        {
+            return MemoryMarshal.CreateSpan(ref UnguardedAccess(span, begin), end - begin);
+        }
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs:1931]: Using unsafe code throughout a complex algorithm like sort is a major liability. The existing sort is intentionally using safe bounds checked access, except a few carefully audited places.

**Comment by jkotas** [comment]
Changing the unstable Array.Sort algorithm is semi-breaking change. It generates a ton of follow up work in projects out there in updating their test baselines. It would be best to switch to a stable sort algorithm that we can keep tweaking for performance in non-breaking way.

What's the best stable sort algorithm available and how does it compare with this and with what we have today? I understand that we would be leaving some perf on the table with stable sort algorithm, but we would be gaining predictability that has a lot of value.

**Comment by hez2010** [comment]
> What's the best stable sort algorithm available and how does it compare with this and with what we have today?

How about this one? https://github.com/scandum/quadsort
From its description in README.md, it even out performs than pdqsort in most cases:

image

**Comment by hez2010** [comment]
> Changing the unstable Array.Sort algorithm is semi-breaking change.

I'm making this PR as draft, and will make it ready-for-review again after changing the implementation to fluxsort.

**Comment by tfenise** [comment]
It seems [fluxsort](https://github.com/scandum/fluxsort) allocates, which may be undesirable in certain cases.

**Comment by danmoseley** [comment]
> stable sort algorithm that we can keep tweaking for performance in non-breaking way.

Subsequently changing to an even better stable algorithm could break baselines again, right?

> quadsort

Just curious, are there algorithms that take advantage of vectorization? I don't see it mentioned there.

**Comment by hez2010** [comment]
> It seems [fluxsort](https://github.com/scandum/fluxsort) allocates, which may be undesirable in certain cases.

Then I can just implement the quadsort, which can be allocation-free via sorting in-place through rotations, but I would need to verify the performance under this circumstance first.

**Comment by hez2010** [comment]
> Changing the unstable Array.Sort algorithm is semi-breaking change.

Aren't we already using an unstable algorithm? We used introsort before which is already unstable.

**Comment by am11** [comment]
> Just curious, are there algorithms that take advantage of vectorization? I don't see it mentioned there.

Looks like in Julia lang, [chipsort](https://github.com/nlw0/ChipSort.jl/blob/64929f5da88893e95dd0f1447484b4c484d4360e/docs/src/theory.md) is implemented based on this research paper:
> [SIMD- and Cache-Friendly Algorithm for Sorting an Array of Structures](http://www.vldb.org/pvldb/vol8/p1274-inoue.pdf), Hiroshi Inoue and Kenjiro Taura (2015)

(among others)

**Comment by hez2010** [comment]
@jkotas Since we already used an unstable sort algorithm, I would like to land this PR (pdqsort).

**Comment by stephentoub** [comment]
> Since we already used an unstable sort algorithm, I would like to land this PR (pdqsort)

It's unstable but determistic. @jkotas'  concern is that a lot of code, especially tests, has taken a dependency on the ordering, and we don't want to churn it only to churn it again.

**Comment by jkotas** [comment]
> Subsequently changing to an even better stable algorithm could break baselines again, right?

No. All stable sort algorithms produce same results. If we switch to a stable algorithm, switching to even better stable algorithm would not change observable behavior.

> Just curious, are there algorithms that take advantage of vectorization? I don't see it mentioned there.

Yes, you can write specialized sort algorithms that take advantage of vectorization and parallelization. They are good candidates for standalone nuget package, not for the good-enough sort algorithm in the core library.

**Comment by hez2010** [comment]
I would like to see the CI first. If there is no failure related to this change, I think it's okay to ship it.
For reference, Rust is also using pdqsort, and Go switched to pdqsort a while ago.

**Comment by hez2010** [comment]
I see CircleInConvex_r and CircleInConvex_ro are failing, but I think the cmp function is malformed:

```csharp
static int cmp(Point a, Point b)
{
    if (a.X < b.X || a.X == b.X && a.Y < b.Y)
        return 1;
    else
        return 0;
}
```

The above cmp function results in indeterminate behavior.

**Comment by danmoseley** [comment]
> No. All stable sort algorithms produce same results. If we switch to a stable algorithm, switching to even better stable algorithm would not change observable behavior.

Ah yes, I was thinking that stable means, always orders elements that compare identically in some predictable order. But actually means, always orders elements that compare identically in the same order they were provided.

File: src/tests/JIT/SIMD/CircleInConvex.cs [hunk 2]
```diff
@@ -258,8 +258,8 @@ static int Main(string[] args)
             float r;
             FindCircle(points, out O, out r);
 
-            float expRes = 75656240.0F;
```
Comment (hez2010) [inline src/tests/JIT/SIMD/CircleInConvex.cs:261]: The original `expRes` is not correct with given points.

**Comment by hez2010** [comment]
Test failures are unrelated.

**Comment by jkotas** [comment]
CircleInConvex is a good test fix. It may be good to submit it as a separate PR.

> If there is no test failure related to this change, I think it's okay to ship it.

I will leave it for @dotnet/area-system-runtime owners to figure out how to drive this behavior change through.

For example, based on past experience - this sort algorithm change would generate a bunch of work for [Bing](https://devblogs.microsoft.com/dotnet/bing-com-runs-on-net-core-2-1/).

I think we can get people behind something like this: "We are sorry for changing the sort algorithm again. This is the last time. The sort algorithm is going to be stable from now and improvements are not going to have observable impact going forward (as long as your compare function is correct).". 

On other hand, "We are sorry for changing the sort algorithm again and we are likely going to do it at least one more time soon so prepare to do the work on your test baselines twice." will be harder to get people behind. There are likely going to be asks for keeping the current sort algorithm under a switch, but that is not particularly great for small binary sizes that are important for other form factors. The whole thing gets complicated.

**Comment by hez2010** [comment]
> It may be good to submit it as a separate PR.

#68475

**Comment by hez2010** [comment]
Most methods of pdqsort can be shared with introsort (insertion sort and heap sort), I think we can provide a runtime switch to fallback to the old introsort behavior, and it won't introduce observable code size regression. 
I think we should land this PR as soon as possible to gather more feedbacks during preview stage. 
@jkotas what do you think?

**Comment by bencyoung-Fignum** [comment]
Does stability matter for native types anyway as you can't tell the difference between two ints with the same value? pdqsort could be used for a known list of types?

**Comment by bencyoung-Fignum** [comment]
Timsort is a good stable sort for references, so you could have Timsort for general sorting and Pdqsort for primitives where stability doesn't matter...

**Comment by GSPP** [comment]
When the algorithm becomes stable, people will take a dependency on that behavior (justified or not). This means that it can never again be made unstable. It's a one-way street.

This feels like .NET 1.0 locking in a specific `Random` algorithm. This 1.0 algorithm was not very good compared to modern standards.

It's also like the .NET 1.0 `String.GetHashCode()`. Nobody should take a dependency on it, but people do. And it makes useful changes that much harder.

**Comment by bencyoung-Fignum** [comment]
> 

I'm just trying to get clarity on what an unstable sort means on primitive structs seeing as you can't tell the difference between two integers with the same value? I can certainly see the difference if you're using an extracted key to sort different values but using an unstable sort for Array.Sort<int[]>(array) seems safe from the stability perspective as you can't tell....

Whether it's worth having a faster unstable sort for this kind of case is another question!

**Comment by hez2010** [comment]
I'm agreeing with @GSPP. Once we switch to a stable sort algorithm and document the behavior, we will never be able to go back to the unstable world again because people will start to rely on the stable behavior.

**Comment by tannergooding** [comment]
That is an intentional and desirable option. Developers are already reliant on the "unstable" sort having not changed and therefore being viewed as "stable" when in actuality its simply deterministic.

**Comment by hez2010** [comment]
Then we can consider adding a `StableSort` method to provide stable sort behavior, so developers can utilize the `StableSort` method for sequence-sensitive jobs.

**Comment by tannergooding** [comment]
As per the above, if any break is taken we are more interested in taking a 1-time break to make the default sorting algorithm stable. This allows us to improve perf over time without continuously hurting developers who may be dependent on a particular behavior.

A faster but unstable sorting algorithm could then be discussed independently, but I imagine its better for that to simply be provided via a 3rd party package maintained by the community.

**Comment by tannergooding** [comment]
> Unless the app provides an IComparer that treats some values as equivalent.

Sure, but that's not comparing primitives, that's using a custom comparer and is effectively treated as "custom type" at that point.

**Comment by jeffhandley** [comment]
Due to the lack of consensus for moving this forward, I'm closing the PR. I will leave #68325 open though and discussion can continue there for under what conditions we would approve a PR for the change. Thanks for the effort on this, @hez2010.

**Pull request closed without merging**